# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"

df = pd.read_csv(url)

print(f"Dimenstions: {df.shape}")
print(df.dtypes)


Dimenstions: (4008, 11)
Customer                         str
ST                               str
GENDER                           str
Education                        str
Customer Lifetime Value          str
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints        str
Policy Type                      str
Vehicle Class                    str
Total Claim Amount           float64
dtype: object


In [3]:
# Retorna una tupla (filas, columnas)
print(f"Dimensiones: {df.shape}")
print(f"Filas: {df.shape[0]}, Columnas: {df.shape[1]}")

Dimensiones: (4008, 11)
Filas: 4008, Columnas: 11


the data type of the number of complaints is inappropriate for the nature of the variable. I suggests to put it in float

In [4]:
# Valores únicos por columna
print(df.nunique())

# Identificación de categóricas
# Una regla común: si tiene pocos valores únicos, suele ser categórica
for col in df.columns:
    if df[col].dtype == 'object' or df[col].nunique() < 15:
        print(f"The Column '{col}' is categorical.")
        print(f"Unique Values: {df[col].unique()}\n")
    else:
        print(f"The Column '{col}' is numerical.")
        print(f"Range: {df[col].min()} to {df[col].max()}\n")

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64
The Column 'Customer' is numerical.
Range: AA71604 to ZZ97035

The Column 'ST' is categorical.
Unique Values: <StringArray>
['Washington',    'Arizona',     'Nevada', 'California',     'Oregon',
       'Cali',         'AZ',         'WA',          nan]
Length: 9, dtype: str

The Column 'GENDER' is categorical.
Unique Values: <StringArray>
[nan, 'F', 'M', 'Femal', 'Male', 'female']
Length: 6, dtype: str

The Column 'Education' is categorical.
Unique Values: <StringArray>
[              'Master',             'Bachelor', 'High School or Below',
              'College',            'Bachelors',               'Do

Data Exploration and Descriptive Insights
Analysis of Customer Demographics and Claim Patterns
1. Data Integrity and Quality Observations
Categorical Fragmentation: High inconsistency is present in the ST (State) and GENDER columns. Labels like 'Cali' vs. 'California' and 'Femal'/'female' vs. 'F' artificially inflate the number of unique values. This indicates a critical need for Data Standardization before any predictive modeling.

Identity vs. Magnitude: Although the Customer column is identified as numerical in the summary, its range (AA71604 to ZZ97035) confirms it is an Alphanumeric Identifier. It should be treated as a unique key and excluded from statistical calculations like means or sums.

Missing Data: The presence of NaN values in geographic and demographic fields suggests a gap in data collection that requires imputation strategies to avoid biased results.

2. Strategic Business Insights
CLV Heterogeneity: With 1,027 unique values for Customer Lifetime Value, the dataset shows a highly diverse customer base. This high variability is an opportunity for Micro-segmentation, allowing the company to tailor loyalty programs to specific value brackets.

Standardized Pricing: The Monthly Premium Auto column has significantly fewer unique values (132) than the total customer count. This suggests that the company uses fixed-tier pricing models rather than hyper-personalized dynamic pricing.

Operational Simplicity: The limited variety in Policy Type (only 3 unique values) indicates a simplified product portfolio. This is ideal for high-level profitability comparisons between Corporate, Personal, and Special lines of business.

3. Risk and Claims Analysis
Claim Severity: The range for Total Claim Amount is vast, stretching from $0.38 to nearly $2,900. This "Heavy-Tail" distribution is typical in the insurance industry, where a small percentage of catastrophic claims accounts for the majority of total costs.

Customer Friction: The Number of Open Complaints is a low-cardinality field (6 values). Analyzing the correlation between these complaints and Response could provide a definitive Churn Prediction metric.

Final Conclusion
"The primary opportunity for the organization lies in Data Governance. By resolving the categorical inconsistencies in gender and geography, the company can move from a descriptive view to a Data-Driven Safety Analysis, accurately identifying which specific segments are driving the high-value claims ($2,900 range)."

Analyst's Note: If you plan to present this, I recommend a Box Plot for Total Claim Amount to visually demonstrate how those extreme values (outliers) are distributed across the different Policy Types.

In [5]:
import pandas as pd
import numpy as np

def perform_statistical_analysis(url: str) -> pd.DataFrame:
    """
    Carga un conjunto de datos y calcula estadísticas descriptivas clave 
    para las columnas numéricas.
    """
    # 1. Carga del dataset
    df = pd.read_csv(url)
    
    # 2. Selección de columnas numéricas
    # Excluimos columnas que actúan como ID (como 'Unnamed: 0' o 'Customer')
    numerical_cols = df.select_dtypes(include=[np.number]).columns.tolist()
    if 'Unnamed: 0' in numerical_cols:
        numerical_cols.remove('Unnamed: 0')

    # 3. Cálculo de estadísticas
    # .describe() nos da media, desviación estándar, cuartiles, min y max
    stats = df[numerical_cols].describe().T
    
    # 4. Añadimos la Mediana y la Moda (que no vienen por defecto en .describe)
    stats['median'] = df[numerical_cols].median()
    
    # La moda puede devolver múltiples valores, tomamos el primero
    stats['mode'] = df[numerical_cols].mode().iloc[0]
    
    # Reordenamos las columnas para mayor claridad
    columns_order = ['mean', 'median', 'mode', 'std', 'min', '25%', '50%', '75%', 'max']
    return stats[columns_order]
stats_results = perform_statistical_analysis("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")
print(stats_results)

                              mean        median   mode           std  \
Income                39295.701214  36234.000000    0.0  30469.427060   
Monthly Premium Auto    193.234360     83.000000   65.0   1601.190369   
Total Claim Amount      404.986909    354.729129  321.6    293.027260   

                            min           25%           50%      75%  \
Income                 0.000000  14072.000000  36234.000000  64631.0   
Monthly Premium Auto  61.000000     68.000000     83.000000    109.5   
Total Claim Amount     0.382107    202.157702    354.729129    532.8   

                               max  
Income                99960.000000  
Monthly Premium Auto  35354.000000  
Total Claim Amount     2893.239678  


Statistical Distribution & Business Insights
1. Income: The Zero-Value Segment
The Mode Trap: The Mode is 0.0, which is a critical finding. This suggests a significant portion of the customer base is likely unemployed or has non-disclosed income.

Income Skewness: The Mean ($39,295) is higher than the Median ($36,234), indicating a right-skewed distribution. High-earning outliers (up to $99,960) are pulling the average up, but the "typical" customer earns closer to $36k.

Strategic Gap: There is a massive gap between the 25th percentile ($14,072) and the maximum. The company is catering to two very different worlds: low-income/unemployed individuals and high-net-worth professionals.

2. Monthly Premium Auto: Extreme Outliers
The "Luxury" Outlier: The Maximum Premium ($35,354) is astronomical compared to the Median ($83). This is likely an error in the data or a very specific high-value commercial fleet policy.

Standard Deviation Volatility: The Std Dev ($1,601) is much higher than the Mean ($193). In a normal distribution, this would be impossible; here, it confirms that the extreme outliers are heavily distorting the "average" cost of a policy.

Standard Market: 75% of your customers pay $109.50 or less. This shows the core business is volume-based, focusing on affordable, standard insurance.

3. Total Claim Amount: Risk Concentration
Predictable Claims: Unlike premiums, claims are more "stable." The Median ($354) and Mean ($404) are relatively close, meaning the distribution is less distorted by extreme values.

The Profitability Buffer: The 75th percentile for claims is $532, while 75% of premiums are only $109. This highlights the insurance fundamental: it takes roughly 5 safe customers paying their monthly premium to cover the cost of 1 average claim.

Maximum Loss: Even the worst-case claim in this set ($2,893) is manageable, as it doesn't reach the extreme levels seen in the premium outliers.

Final Executive Summary
"The data reveals a stable core business (Medians: $36k Income / $83 Premium) but identifies extreme volatility in Premium pricing that requires investigation. The high frequency of zero-income customers (Mode: 0) suggests a need for targeted products for the unemployed segment, while the claim distribution confirms that the company's financial risk is well-contained within the $2,900 threshold."

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [6]:
import pandas as pd

# 1. Create the Series with frequencies
# We use value_counts() which returns a Series indexed by location
location_counts = df['ST'].value_counts()

# 2. Retrieve the top 5 less common locations
# We sort in ascending order (default) to get the smallest counts first
bottom_5_locations = location_counts.sort_values().head(5)

# Display the result
print("Top 5 less common customer locations:")
print(bottom_5_locations)


Top 5 less common customer locations:
ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [7]:
import pandas as pd

# 1. Create a pandas Series with the counts of each policy type
policy_counts = df['Policy Type'].value_counts()

# 2. Retrieve the policy type with the highest number of policies sold
# Since value_counts() sorts by descending order, the first index [0] is the highest
highest_policy_type = policy_counts.idxmax()
highest_value = policy_counts.max()

# Display the results
print("Total policies sold by type:")
print(policy_counts)
print(f"\nThe policy type with the highest sales is: {highest_policy_type} ({highest_value} policies)")

Total policies sold by type:
Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

The policy type with the highest sales is: Personal Auto (780 policies)


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [8]:
# Your code hereimport pandas as pd

# 1. Filtramos los datos para obtener solo las filas de 'Personal Auto'
# Luego seleccionamos la columna 'Income'
personal_auto_income = df.loc[df['Policy Type'] == 'Personal Auto', 'Income']

# 2. Filtramos los datos para obtener solo las filas de 'Corporate Auto'
# Luego seleccionamos la columna 'Income'
corporate_auto_income = df.loc[df['Policy Type'] == 'Corporate Auto', 'Income']

# 3. Calculamos el promedio (mean) de cada grupo por separado
mean_personal = personal_auto_income.mean()
mean_corporate = corporate_auto_income.mean()

# 4. Mostramos los resultados
print(f"Average Income (Personal Auto):  ${mean_personal:,.2f}")
print(f"Average Income (Corporate Auto): ${mean_corporate:,.2f}")


Average Income (Personal Auto):  $38,180.70
Average Income (Corporate Auto): $41,390.31


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code herecc